### PR TITLE
Compound hets

### PR DESCRIPTION
**Pull request by lizzij**
Compound Hets
- parse compound het pairs from elastic search results
- filter by optional second hit annotation filter in recessive searches
  - second hit accordion popup for recessive searches
- displaying compound hets in pairs
  - new ui (pagination, link, shared notes, tags, functional data section)
  - db migration 1 : 1 -> many : many for variant : notes/tags/functional data
  - crud for shared/singular variant notes and tags
  - display saved variants in compound het pairs
  - toggle pair/unpair
- tests

### CODE REVIEW

File: seqr/migrations/0064_auto_20190826_1748.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        migrations.RemoveField(
+            model_name='variantnote',
+            name='saved_variant',
+        ),
+        migrations.AddField(
+            model_name='variantnote',
+            name='saved_variant',
+            field=models.ManyToManyField(to='seqr.SavedVariant'),
+        ),
+        migrations.RemoveField(
```
Comment (hanars) [inline seqr/migrations/0064_auto_20190826_1748.py]: this almost definitely effectively deletes all the tags/ notes in seqr. Lets discuss in person

File: seqr/models.py [hunk 2]
```diff
@@ -608,19 +608,18 @@ class VariantTag(ModelWithGUID):
     search_parameters = models.TextField(null=True, blank=True)  # aka. search url
 
     def __unicode__(self):
-        return "%s:%s" % (str(self.saved_variant), self.variant_tag_type.name)
+        return "%s:%s" % (str(self.saved_variants), self.variant_tag_type.name)
```
Comment (hanars) [inline seqr/models.py]: does this work or does this create something with weird brackets in the guid or something?

File: seqr/utils/es_utils.py [hunk 3]
```diff
@@ -70,8 +70,8 @@ def process_previous_results(previous_search_results):
             num_results_to_use = total_results or 10000
         start_index = (page - 1) * num_results_to_use
         end_index = page * num_results_to_use
-        if previous_search_results.get('total_results') is not None:
-            end_index = min(end_index, previous_search_results['total_results'])
+        if total_results is not None:
+            end_index = min(end_index, total_results)
```
Comment (hanars) [inline seqr/utils/es_utils.py]: merge issue?

File: seqr/utils/es_utils.py [hunk 4]
```diff
@@ -564,6 +564,7 @@ def _parse_compound_het_response(self, response):
                 for family_guid in variant['familyGuids']:
                     family_variants[family_guid].append(variant)
 
+            # TODO place compound hets from the same family together
```
Comment (hanars) [inline seqr/utils/es_utils.py]: this seems important?

File: seqr/views/apis/saved_variant_api.py [hunk 5]
```diff
@@ -72,110 +61,180 @@ def create_saved_variant_handler(request):
...
+def create_saved_variant_handler(request):
+    variant_json = json.loads(request.body)
+    non_variant_json = {
+        k: variant_json.pop(k, None) for k in ['searchHash', 'tags', 'functionalData', 'notes', 'note', 'submitToClinvar']
+    }
+    family_guid = variant_json.pop('familyGuid')
+    family = Family.objects.get(guid=family_guid)
+    check_permissions(family.project, request.user, CAN_VIEW)
+    # single gene
+    if 'familyGuids' in variant_json.keys():
```
Comment (hanars) [inline seqr/views/apis/saved_variant_api.py]: this is a somewhat confusing way to check if its a compound het or not. Is there a more explicit check you could do, like what data type variant json is?

File: seqr/views/apis/saved_variant_api.py [hunk 6]
```diff
@@ -72,110 +61,180 @@ def create_saved_variant_handler(request):
...
+        elif non_variant_json.get('tags'):
+            _create_new_tags([saved_variant], non_variant_json, request.user)
+        variant_json.update(get_json_for_saved_variant(saved_variant, add_tags=True))
+        return create_json_response({
+            'savedVariantsByGuid': {saved_variant.guid: variant_json},
+        })
+
+    # compound hets
+    else:
+        compound_hets_json = variant_json
```
Comment (hanars) [inline seqr/views/apis/saved_variant_api.py]: it looks like there is a bunch of copy-paste between this code and the single variant code. Maybe make a helper function?

File: seqr/views/apis/saved_variant_api.py [hunk 7]
```diff
@@ -72,110 +61,180 @@ def create_saved_variant_handler(request):
...
 
 @login_required(login_url=API_LOGIN_REQUIRED_URL)
 @csrf_exempt
-def delete_variant_note_handler(request, variant_guid, note_guid):
-    saved_variant = SavedVariant.objects.get(guid=variant_guid)
-    check_permissions(saved_variant.family.project, request.user, CAN_VIEW)
-    note = VariantNote.objects.get(guid=note_guid, saved_variant=saved_variant)
+def delete_variant_note_handler(request, variant_guids, note_guid):
+    variant_guids = variant_guids.split(',')
+    saved_variants = []
```
Comment (hanars) [inline seqr/views/apis/saved_variant_api.py]: You can do this as one query with `saved_variants = SavedVariant.objects.get(guid__in= variant_guids)`

File: seqr/views/apis/saved_variant_api.py [hunk 8]
```diff
@@ -72,110 +61,180 @@ def create_saved_variant_handler(request):
...
+        saved_variants.append(saved_variant)
+    note = VariantNote.objects.get(guid=note_guid)
     delete_seqr_model(note)
-    return create_json_response({'savedVariantsByGuid': {variant_guid: {
-        'notes': [get_json_for_variant_note(tag) for tag in saved_variant.variantnote_set.all()]
-    }}})
+    update = {}
+    for variant_guid in variant_guids:
+        update[variant_guid] = {
+            'notes': [get_json_for_variant_note(note) for note in saved_variant.variantnote_set.all()]
```
Comment (hanars) [inline seqr/views/apis/saved_variant_api.py]: what is `saved_variant` here?

File: seqr/views/apis/saved_variant_api.py [hunk 9]
```diff
@@ -72,110 +61,180 @@ def create_saved_variant_handler(request):
...
-    saved_variant = SavedVariant.objects.get(guid=variant_guid)
-    check_permissions(saved_variant.family.project, request.user, CAN_VIEW)
-    note = VariantNote.objects.get(guid=note_guid, saved_variant=saved_variant)
+def update_variant_note_handler(request, variant_guids, note_guid):
+    variant_guids = variant_guids.split(',')
+    saved_variants = []
+    for variant_guid in variant_guids:
+        saved_variant = SavedVariant.objects.get(guid=variant_guid)
+        check_permissions(saved_variant.family.project, request.user, CAN_VIEW)
+        saved_variants.append(saved_variant)
```
Comment (hanars) [inline seqr/views/apis/saved_variant_api.py]: You can do this as one query with saved_variants = SavedVariant.objects.get(guid__in= variant_guids)

File: seqr/views/apis/saved_variant_api.py [hunk 10]
```diff
@@ -72,110 +61,180 @@ def create_saved_variant_handler(request):
...
     request_json = json.loads(request.body)
     update_model_from_json(note, request_json, allow_unknown_keys=True)
 
-    return create_json_response({'savedVariantsByGuid': {variant_guid: {
-        'notes': [get_json_for_variant_note(tag) for tag in saved_variant.variantnote_set.all()]
-    }}})
+    update = {}
+    for variant_guid in variant_guids:
+        update[variant_guid] = {
+            'notes': [get_json_for_variant_note(note) for note in saved_variant.variantnote_set.all()],
```
Comment (hanars) [inline seqr/views/apis/saved_variant_api.py]: what is saved_variant here?

File: seqr/views/apis/saved_variant_api.py [hunk 11]
```diff
@@ -189,43 +248,48 @@ def update_variant_tags_handler(request, variant_guid):
             tag_model = VariantFunctionalData.objects.get(
                 guid=tag.get('tagGuid'),
                 functional_data_tag=tag.get('name'),
-                saved_variant=saved_variant
+                saved_variants=saved_variants
```
Comment (hanars) [inline seqr/views/apis/saved_variant_api.py]: does this work?

File: seqr/views/apis/saved_variant_api.py [hunk 12]
```diff
@@ -189,43 +248,48 @@ def update_variant_tags_handler(request, variant_guid):
...
+    update = {}
+    for variant_guid in variant_guids:
+        update[variant_guid] = {
             'tags': [get_json_for_variant_tag(tag) for tag in saved_variant.varianttag_set.all()],
             'functionalData': [get_json_for_variant_functional_data(tag) for tag in saved_variant.variantfunctionaldata_set.all()]
         }
-    }})
+
+    import pdb
+    pdb.set_trace()
```
Comment (hanars) [inline seqr/views/apis/saved_variant_api.py]: lol

File: seqr/views/apis/variant_search_api.py [hunk 13]
```diff
@@ -469,45 +485,39 @@ def _get_saved_searches(user):
...
         #  For saved variants only use family it was saved for, not all families in search
         saved_variant['familyGuids'] = family_guids
         saved_variants_by_guid[saved_variant['variantGuid']] = saved_variant
-        if searched_variant['variantId'] not in variants_to_saved_variants:
-            variants_to_saved_variants[searched_variant['variantId']] = {}
-        for family_guid in family_guids:
-            variants_to_saved_variants[searched_variant['variantId']][family_guid] = saved_variant['variantGuid']
 
-    return saved_variants_by_guid, variants_to_saved_variants
+    return saved_variants_by_guid
```
Comment (hanars) [inline seqr/views/apis/variant_search_api.py]: is this a merge issue? That change is pretty important...

File: seqr/views/apis/variant_search_api.py [hunk 14]
```diff
@@ -469,45 +485,39 @@ def _get_saved_searches(user):
...
 
 
-def _get_variant_key(xpos=None, ref=None, alt=None, genomeVersion=None, **kwargs):
```
Comment (hanars) [inline seqr/views/apis/variant_search_api.py]: merge issue?

File: seqr/views/utils/orm_to_json_utils.py [hunk 15]
```diff
@@ -576,6 +576,41 @@ def _get_collaborator_json(collaborator, include_permissions, can_edit):
         })
     return collaborator_json
 
+#
+# def get_project_variant_tag_types(project, tag_counts_by_type_and_family=None, note_counts_by_family=None):
```
Comment (hanars) [inline seqr/views/utils/orm_to_json_utils.py]: probably should take this out

File: ui/pages/Project/components/FamilyPage.jsx [hunk 16]
```diff
@@ -9,24 +9,27 @@ import Family from 'shared/components/panel/family'
...
-      family={family}
-      compact={compact}
-      {...props}
-    />
-    {!compact && <ShowReadsButton familyGuid={family.familyGuid} igvId={family.familyGuid} padding="0.5em 0 1.5em 0" />}
-    <FamilyVariantReads igvId={family.familyGuid} />
-    {individuals && individuals.map(individual => (
-      <IndividualRow
-        key={individual.individualGuid}
+const BaseFamilyDetail = ({ family, individuals, editCaseReview, compact, ...props }) => {
```
Comment (hanars) [inline ui/pages/Project/components/FamilyPage.jsx]: why did this change?

File: ui/redux/rootReducer.js [hunk 17]
```diff
@@ -277,15 +277,52 @@ const updateSavedVariant = (values, action = 'create') => {
 }
 
 export const updateVariantNote = (values) => {
-  if (values.variantGuid) {
-    return updateEntity(values, RECEIVE_DATA, `/api/saved_variant/${values.variantGuid}/note`, 'noteGuid')
+  if (values.familyGuids) {
+    if (values.variantGuid) {
```
Comment (hanars) [inline ui/redux/rootReducer.js]: why add the familyGuids check?

File: ui/redux/rootReducer.js [hunk 18]
```diff
@@ -277,15 +277,52 @@ const updateSavedVariant = (values, action = 'create') => {
...
+      return updateEntity(values, RECEIVE_DATA, `/api/saved_variant/${values.variantGuid}/note`, 'noteGuid')
+    }
+    return updateSavedVariant(values)
+  }
+  if (values[0].variantGuid) {
+    const variantGuids = []
+    let currVariantGuid
+    for (let i = 0; i < Object.keys(values).length; i++) {
+      try {
+        currVariantGuid = values[i].variantGuid
```
Comment (hanars) [inline ui/redux/rootReducer.js]: Isn't `values` and array of objects? Couldn't this just be `values.map(({ variantGuid }) => variantGuid).filter(variantGuid => variantGuid)`

File: ui/shared/components/panel/variants/FamilyVariantTags.jsx [hunk 19]
```diff
@@ -205,124 +216,150 @@ const VariantNoteField = ({ action, note, variant, family, ...props }) => {
...
-            field="tags"
-            fieldName="Tags"
-            family={family}
-            variant={savedVariant || variant}
-            tagOptions={project.variantTagTypes.filter(vtt => vtt.name !== NOTE_TAG_NAME)}
-            onSubmit={dispatchUpdateFamilyVariantTags}
-          />
-          <HorizontalSpacer width={5} />
-          {savedVariant && savedVariant.tags.some(tag => tag.category === DISCOVERY_CATEGORY_NAME) &&
+  { variant, savedVariant, family, project, dispatchUpdateVariantNote, dispatchUpdateFamilyVariantTags, isCompoundHet, areCompoundHets },
```
Comment (hanars) [inline ui/shared/components/panel/variants/FamilyVariantTags.jsx]: what is the difference between `isCompoundHet` and `areCompoundHets`?

File: ui/shared/components/panel/variants/Variants.jsx [hunk 20]
```diff
@@ -68,14 +86,76 @@ const Variant = ({ variant }) => {
...
+    </StyledCompoundHetRow>
+  )
+}
+
+CompoundHet.propTypes = {
+  variant: PropTypes.object,
+}
+
+const CompoundHets = ({ variants }) => {
+  const geneId = getVariantMainGeneId(variants[0])
```
Comment (hanars) [inline ui/shared/components/panel/variants/Variants.jsx]: you actually want to use the gene shared between the variants, which isn't always the main gene for the first variant

File: ui/shared/components/panel/variants/Variants.jsx [hunk 21]
```diff
@@ -68,14 +86,76 @@ const Variant = ({ variant }) => {
   )
 }
 
-Variant.propTypes = {
+SingleVariant.propTypes = {
   variant: PropTypes.object,
 }
 
+const CompoundHet = ({ variant }) => {
```
Comment (hanars) [inline ui/shared/components/panel/variants/Variants.jsx]: this is almost a copy-paste of `SingleVariant` so you should modify that component to work for both (maybe add a `showGeneInfo` prop that is true for regular variants and false for compound hets or something like that

File: ui/shared/components/panel/variants/FamilyVariantTags.jsx (hunk 19 above)
Comment (lizzij) [reply ui/shared/components/panel/variants/FamilyVariantTags.jsx]: `isCompoundHet` is for checking if a single variant is compound het, while `areCompoundHets` is for checking if a list of variants are compound hets (to style the main gene section).

File: seqr/utils/es_utils.py (hunk 4 above)
Comment (lizzij) [reply seqr/utils/es_utils.py]: `.append(variant)` seems to implicitly ensure that compound hets from the same family are placed together.

File: seqr/views/apis/variant_search_api.py (hunk 14 above)
Comment (lizzij) [reply seqr/views/apis/variant_search_api.py]: Is fixed now.

File: seqr/utils/es_utils.py (hunk 4 above)
Comment (lizzij) [reply seqr/utils/es_utils.py]: Need to design a new UI to label compound hets from different families.

File: ui/redux/rootReducer.js (hunk 17 above)
Comment (lizzij) [reply ui/redux/rootReducer.js]: to check if it is a single variant

File: seqr/utils/es_utils.py (hunk 4 above)
Comment (lizzij) [reply seqr/utils/es_utils.py]: ^ Display in pairs

File: ui/shared/components/panel/variants/FamilyVariantTags.jsx [hunk 22]
```diff
@@ -205,124 +221,150 @@ const VariantNoteField = ({ action, note, variant, family, ...props }) => {
...
   family: PropTypes.object.isRequired,
 }
 
+const VariantLink = (
+  { variant, savedVariant, family },
+) =>
+  <VariantLinkContainer>
+    <NavLink
+      to={Array.isArray(savedVariant) ?
+        `/variant_search/variant/${variant.variantId}/family/${family.familyGuid}` :
```
Comment (hanars) [inline ui/shared/components/panel/variants/FamilyVariantTags.jsx]: this needs to have a link to a saved variants page, not a variant search page

File: ui/shared/components/panel/variants/Variants.jsx [hunk 23]
```diff
@@ -12,13 +12,15 @@ import Predictions from './Predictions'
...
-  .column {
-    margin-top: 1em !important;
-    margin-bottom: 0 !important;
-  }
+const StyledVariantRow = styled(Grid.Row)`  
+  ${({ iscompoundhet }) => {
+    return (iscompoundhet ?
+      '.column { margin-top: 0em !important; margin-left: 1em !important; }' :
+      '.column { margin-top: 1em !important; margin-bottom: 0 !important; margin-left: 1em !important; }')
+  }}
```
Comment (hanars) [inline ui/shared/components/panel/variants/Variants.jsx]: You should do the conditional logic inside the `.column` block like:
```
.column {
   ${({ iscompoundhet }) =>  iscompoundhet ? { margin-top: 0em !important; margin-left: 1em !important; } : ...
}
```

File: ui/shared/components/panel/variants/Variants.jsx [hunk 24]
```diff
@@ -34,33 +36,63 @@ const StyledVariantRow = styled(Grid.Row)`
...
+  margin-right: 1em !important;
+  margin-top: 0.5em !important;
+  margin-bottom: 0 !important;
+`
+
+const Variant = ({ variant, isCompoundHet }) => {
   const mainGeneId = getVariantMainGeneId(variant)
   return (
-    <StyledVariantRow severity={CLINSIG_SEVERITY[(variant.clinvar.clinicalSignificance || '').toLowerCase()]}>
+    <StyledVariantRow key={variant.variantId} severity={CLINSIG_SEVERITY[(variant.clinvar.clinicalSignificance || '').toLowerCase()]} iscompoundhet={(isCompoundHet || false).toString()} >
```
Comment (hanars) [inline ui/shared/components/panel/variants/Variants.jsx]: why are you using toString?

File: ui/shared/components/panel/variants/Variants.jsx [hunk 25]
```diff
@@ -34,33 +36,63 @@ const StyledVariantRow = styled(Grid.Row)`
...
           <FamilyVariantTags familyGuid={familyGuid} variant={variant} />
         </Grid.Column>,
       )}
+      {isCompoundHet &&
+      <Grid.Column>
+        {Object.keys(variant.transcripts).filter(geneId => geneId !== mainGeneId).map(geneId =>
+          <VariantGene key={geneId} geneId={geneId} variant={variant} compact />,
+        )}
+        {Object.keys(variant.transcripts).length > 1 && <VerticalSpacer height={20} />}
+        {variant.familyGuids.map(familyGuid =>
```
Comment (hanars) [inline ui/shared/components/panel/variants/Variants.jsx]: for things like this where you end up copying the same code twice so you can put it in a different place for compound hets you should assign it to a const once at the top and then reference it in multiple places. So you would have `const variantIndividuals = ...` up where `mainGeneId` is defined and then here all you would have is `{variantIndividuals}`

File: ui/shared/components/panel/variants/Variants.jsx [hunk 26]
```diff
@@ -70,12 +102,42 @@ const Variant = ({ variant }) => {
...
+
+CompoundHets.propTypes = {
+  variants: PropTypes.array,
+}
+
 const Variants = ({ variants }) =>
   <Grid stackable divided="vertically" columns="equal">
-    {variants.map(variant => <Variant key={variant.variantId} variant={variant} />)}
+    {variants.map(variant =>
+      (variant.length > 1 ? <CompoundHets variants={variant} key={variant[0].variantId} /> : <Variant variant={variant} key={variant.variantId} />))}
```
Comment (hanars) [inline ui/shared/components/panel/variants/Variants.jsx]: for the compound het key, you need to actually include both variant IDs because you theoretically could end up with multiple pairs of compound hets with the same first variant

File: ui/shared/components/panel/view-fields/BaseFieldView.jsx [hunk 27]
```diff
@@ -156,6 +155,7 @@ BaseFieldView.propTypes = {
   field: PropTypes.string.isRequired,
   idField: PropTypes.string,
   initialValues: PropTypes.object,
+  displayIdField: PropTypes.string,
```
Comment (hanars) [inline ui/shared/components/panel/view-fields/BaseFieldView.jsx]: why is this necessary?

File: ui/shared/utils/constants.js [hunk 28]
```diff
@@ -807,6 +807,13 @@ export const getVariantMainGeneId = ({ transcripts, mainTranscriptId, selectedMa
     entry[1].some(({ transcriptId }) => transcriptId === (selectedMainTranscriptId || mainTranscriptId)),
   ) || [])[0]
 
+export const getCompoundHetsAllGeneIds = compoundHets => compoundHets.map(({ transcripts }) => Object.keys(transcripts))
+
```
Comment (hanars) [inline ui/shared/utils/constants.js]: this will return a list that has duplicate gene IDs

File: ui/shared/utils/constants.js [hunk 29]
```diff
@@ -807,6 +807,13 @@ export const getVariantMainGeneId = ({ transcripts, mainTranscriptId, selectedMa
...
+export const getCompoundHetsSharedGeneIds = geneIds =>
+  (geneIds.shift().filter(sameVariantGeneIds =>
```
Comment (hanars) [inline ui/shared/utils/constants.js]: This is really hard to read. What format are `geneIds` in?

File: xbrowse_server/base/model_utils.py [hunk 30]
```diff
@@ -130,13 +131,13 @@ def _deprecated_get_or_create_saved_variant(xpos=None, ref=None, alt=None, famil
 
 XBROWSE_TO_SEQR_ADDITIONAL_ENTITIES_MAPPING = {
     "VariantTag": {
-        "saved_variant": _deprecated_get_or_create_saved_variant
+        "saved_variants": _deprecated_get_or_create_saved_variant
```
Comment (hanars) [inline xbrowse_server/base/model_utils.py]: make sure to change the return type of `_deprecated_get_or_create_saved_variant` from a single variant to a list

File: ui/shared/components/panel/variants/Variants.jsx (hunk 24 above)
Comment (lizzij) [reply ui/shared/components/panel/variants/Variants.jsx]: I got the following error message when using a boolean:

```
backend.js:6 Warning: Received `true` for a non-boolean attribute `iscompoundhet`.

If you want to write it to the DOM, pass a string instead: iscompoundhet="true" or iscompoundhet={value.toString()}.
   ...
```

File: ui/shared/utils/constants.js (hunk 29 above)
Comment (lizzij) [reply ui/shared/utils/constants.js]: `geneIds` is an array of array of strings, and the method is trying to find the first common string amongst the arrays of strings, i.e., the shared geneId amongst the compound hets.
